core/variable_operations.cpp

```diff
@@ -294,4 +294,88 @@ Variable masks_merge_if_contained(const MasksConstProxy &masks,
+}
+} // namespace
+
+VariableProxy nan_to_num(const VariableConstProxy &var,
+                         const VariableConstProxy &replacement,
+                         const VariableProxy &out) {
+
+  validate_inputs_for_replacement(var, replacement);
+
+  transform_in_place<pair_self_t<double, float>>(
+      out, var,
+      scipp::overloaded{
+          [&](auto &out_v, const auto &in_v) {
+            using std::isnan;
+            using V_OUT = std::decay_t<decltype(out_v)>;
+            using V_IN = std::decay_t<decltype(in_v)>;
+            if constexpr (is_ValueAndVariance_v<V_OUT>) {
+              if constexpr (is_ValueAndVariance_v<V_IN>) {
+                using value_type = typename V_IN::value_type;
+                if (isnan(in_v)) {
+                  out_v.value = replacement.value<value_type>();
+                  out_v.variance = replacement.hasVariances()
+                                       ? replacement.variance<value_type>()
+                                       : out_v.value;
+                } else {
+                  // Assign directly to output
+                  out_v.value = in_v.value;
+                  out_v.variance = in_v.variance;
+                }
+              } else {
+                if (isnan(in_v)) {
+                  out_v.value = replacement.value<V_IN>();
+                  out_v.variance = replacement.hasVariances()
+                                       ? replacement.variance<V_IN>()
+                                       : out_v.value;
+                } else {
+                  // Assign directly to output
+                  out_v.value = in_v;
+                  out_v.variance = in_v;
+                }
+              }
+            } else {
+              out_v = isnan(in_v) ? replacement.value<V_IN>() : in_v;
+            }
+          },
+          [&](units::Unit &ua, const units::Unit &ub) {
+            expect::equals(ua, ub);
+          }});
+  return out;
+}
+
+Variable nan_to_num(const VariableConstProxy &var,
+                    const VariableConstProxy &replacement) {
+  validate_inputs_for_replacement(var, replacement);
+
+  return transform<pair_custom_t<double, float>>(
+      var, replacement,
+      overloaded{[&](const auto &x, const auto &repl) {
+                   using std::isnan;
+                   using V = std::decay_t<decltype(x)>;
+                   using R = std::decay_t<decltype(repl)>;
+
+                   if constexpr (is_ValueAndVariance_v<V>) {
+                     if constexpr (is_ValueAndVariance_v<R>)
+                       return isnan(x) ? repl : x;
+                     else
+                       return isnan(x) ? V{repl, repl} : x;
+                   } else {
+                     if constexpr (is_ValueAndVariance_v<R>)
+                       return isnan(x) ? repl.value : x;
+                     else
+                       return isnan(x) ? repl : x;
+                   }
```

Comment:
I think we should avoid all this branching. How about adding a flag for `transform` to specify that only inputs with matching presence of variances are accepted? Then this lambda is a one-liner.